### PR TITLE
BUG: Return local Timestamp.weekday_name attribute (#17354)

### LINKED ISSUE #17354: BUG: Timestamp.weekday_name uses UTC instead of given timezone
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd

# Create a Timestamp for 11pm Monday 28th
ts = pd.Timestamp('2017-08-28 23:00:00')
# Both weekday() and weekday_name indicate this is a Monday
assert ts.weekday() == 0
assert ts.weekday_name == 'Monday'

# Now localize the timezone
ts = ts.tz_localize('EST')
# weekday() still says Monday, but weekday_name says Tuesday!
assert ts.weekday() == 0
assert ts.weekday_name == 'Tuesday'
```
#### Problem description

For `pd.Timestamp` objects with timezone information, the `weekday_name` attribute appears to give the name of the weekday in UTC (in the above example, this would be 2017-08-29 04:00:00), whereas `weekday()` uses local time. I could find no mention of this difference in the documentation for `weekday` and `weekday_name`. I would expect both to work in the given timezone.

#### Expected Output

```
Expected:
>>> pd.Timestamp('2017-08-28 23:00:00', tz='EST').weekday_name
'Monday'

Actual:
>>> pd.Timestamp('2017-08-28 23:00:00', tz='EST').weekday_name
'Tuesday'
```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.1.final.0
python-bits: 64
OS: Windows
OS-release: 10
machine: AMD64
processor: Intel64 Family 6 Model 58 Stepping 9, GenuineIntel
byteorder: little
LC_ALL: C
LANG: None
LOCALE: None.None

pandas: 0.20.3
pytest: None
pip: 9.0.1
setuptools: 36.2.7
Cython: None
numpy: 1.13.1
scipy: None
xarray: None
IPython: None
sphinx: None
patsy: None
dateutil: 2.6.1
pytz: 2017.2
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: None
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
sqlalchemy: None
pymysql: None
psycopg2: 2.7.3.1 (dt dec pq3 ext lo64)
jinja2: 2.9.6
s3fs: None
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by mroeschke**
- [x] closes #17354
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

@jreback created a new method `_get_named_field` based on your [comment](https://github.com/pandas-dev/pandas/issues/17354#issuecomment-325319229) in the issue thread.

### CODE REVIEW

File: doc/source/whatsnew/v0.21.0.txt [hunk 1]
```diff
@@ -342,6 +342,7 @@ Conversion
 - Fixed the return type of ``IntervalIndex.is_non_overlapping_monotonic`` to be a Python ``bool`` for consistency with similar attributes/methods.  Previously returned a ``numpy.bool_``. (:issue:`17237`)
 - Bug in ``IntervalIndex.is_non_overlapping_monotonic`` when intervals are closed on both sides and overlap at a point (:issue:`16560`)
 - Bug in :func:`Series.fillna` returns frame when ``inplace=True`` and ``value`` is dict (:issue:`16156`)
+- Bug in ``Timestamp.weekday_name`` returning a UTC-based weekday name when localized to a timezone (:issue:`17354`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.21.0.txt]: you can do `` :func:`Timestamp.weekday_name` `` I think (might need to be attr actually)

File: pandas/_libs/tslib.pyx [hunk 2]
```diff
@@ -1268,13 +1266,23 @@ cdef class _Timestamp(datetime):
         # same timezone if specified)
         return datetime.__sub__(self, other)
 
-    cpdef _get_field(self, field):
+    cpdef _maybe_convert_value_to_local(self):
```
Comment (jreback) [inline pandas/_libs/tslib.pyx]: this can be just cdef and should return ``int64_t``

File: pandas/_libs/tslib.pyx [hunk 3]
```diff
@@ -1268,13 +1266,23 @@ cdef class _Timestamp(datetime):
...
+        """Convert UTC i8 value to local i8 value if tz exists"
         val = self.value
         if self.tz is not None and not _is_utc(self.tz):
             val = tz_convert_single(self.value, 'UTC', self.tz)
+        return val
+
+    cpdef _get_field(self, field):
+        val = self._maybe_convert_value_to_local()
```
Comment (jreback) [inline pandas/_libs/tslib.pyx]: you can type val as ``cdef: int64_t`` (same for ``_get_named_field``)

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by mroeschke** [comment]
cdef'd `_maybe_convert_value_to_local` and `val` argument in related function and all green.

File: pandas/_libs/tslib.pyx [hunk 4]
```diff
@@ -1268,13 +1266,27 @@ cdef class _Timestamp(datetime):
         # same timezone if specified)
         return datetime.__sub__(self, other)
 
-    cpdef _get_field(self, field):
+    cdef int64_t _maybe_convert_value_to_local(self):
+        """Convert UTC i8 value to local i8 value if tz exists"""
```
Comment (jreback) [inline pandas/_libs/tslib.pyx]: you need to add a

```
cdef int64_t val
```

here in order to type it

File: pandas/_libs/tslib.pyx [hunk 5]
```diff
@@ -1268,13 +1266,27 @@ cdef class _Timestamp(datetime):
...
         val = self.value
         if self.tz is not None and not _is_utc(self.tz):
             val = tz_convert_single(self.value, 'UTC', self.tz)
+        return val
+
+    cpdef _get_field(self, field):
+        cdef:
```
Comment (jreback) [inline pandas/_libs/tslib.pyx]: like this

File: pandas/_libs/tslib.pyx [hunk 6]
```diff
@@ -1268,13 +1266,27 @@ cdef class _Timestamp(datetime):
...
+            int64_t val
+        val = self._maybe_convert_value_to_local()
```
Comment (jreback) [inline pandas/_libs/tslib.pyx]: if you can type out in these (ndarray[int32_t]) i think

**Review by jreback — CHANGES_REQUESTED** [review]
some cdef comments. lgtm. ping when changed / pushed.

File: pandas/_libs/tslib.pyx [hunk 7]
```diff
@@ -1268,13 +1266,33 @@ cdef class _Timestamp(datetime):
...
             val = tz_convert_single(self.value, 'UTC', self.tz)
-        out = get_date_field(np.array([val], dtype=np.int64), field)
+        return val
+
+    cpdef _get_field(self, field):
+        cdef:
+            int64_t val
+            ndarray[int64_t] date_array = np.empty(1, dtype=np.int64)
+        val = self._maybe_convert_value_to_local()
+        date_array[0] = val
```
Comment (jreback) [inline pandas/_libs/tslib.pyx]: this is overkill to type the input ndarray itself

it was ``out`` which was looking to type, though not a big deal, so just remove the date_array temp. ping on green.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks!

**Pull request merged**